src/algebra/continued_fractions/computation/approximations.lean

```diff
@@ -0,0 +1,278 @@
+        (or.inr not_termianted_at_ppred_n''),
+      suffices : (fib (n'' + 1) : K) ≤ gp.b * pconts.b,
+        solve_by_elim [(add_le_add ppred_nth_fib_le_ppconts_B)],
+      -- finally use the fact that 1 ≤ gp.b to solve the goal
+      suffices : 1 * (fib (n'' + 1) : K) ≤ gp.b * pconts.b, by rwa [one_mul] at this,
+      have one_le_gp_b : (1 : K) ≤ gp.b, from
+        of_one_le_nth_part_denom (part_denom_eq_s_b s_ppred_nth_eq),
+      have : (0 : K) ≤ fib (n'' + 1), by exact_mod_cast (fib (n'' + 1)).zero_le,
+      have : (0 : K) ≤ gp.b, from le_trans zero_le_one one_le_gp_b,
+      mono }}
+end)
+
+/-- Shows that the `n`th denominator is greater or equal than `n + 1`th fibonacci number, that is
+`nat.fib (n + 1) ≤ Bₙ`. -/
+lemma of_succ_nth_fib_le_nth_denom (hyp: n = 0 ∨ ¬(gcf.of v).terminated_at (n - 1)) :
+  (fib (n + 1) : K) ≤ (gcf.of v).denominators n :=
+begin
+  rw [denom_eq_conts_b, nth_cont_eq_succ_nth_cont_aux],
+  have : (n + 1) ≤ 1 ∨ ¬(gcf.of v).terminated_at (n - 1), by
+  { cases n,
+    case nat.zero : { exact (or.inl $ le_refl 1) },
+    case nat.succ : { exact or.inr (or.resolve_left hyp n.succ_ne_zero) }},
+  exact (of_fib_le_continuants_aux_b this)
+end
+
+lemma of_le_succ_succ_nth_continuants_aux_b {b : K}
+(nth_part_denom_eq : (gcf.of v).partial_denominators.nth n = some b) :
+  b * ((gcf.of v).continuants_aux $ n + 1).b ≤ ((gcf.of v).continuants_aux $ n + 2).b :=
+begin
+  set g := gcf.of v with g_eq,
+  obtain ⟨gp_n, nth_s_eq, ⟨refl⟩⟩ : ∃ gp_n, g.s.nth n = some gp_n ∧ gp_n.b = b, from
+    obtain_s_b_of_part_denom nth_part_denom_eq,
+  let conts := g.continuants_aux (n + 2),
+  set pconts := g.continuants_aux (n + 1) with pconts_eq,
+  set ppconts := g.continuants_aux n with ppconts_eq,
+  -- use the recurrence of continuants_aux and the fact that gp_n.a = 1
+  suffices : gp_n.b * pconts.b ≤ ppconts.b + gp_n.b * pconts.b, by
+  { have : gp_n.a = 1, from of_part_num_eq_one (part_num_eq_s_a nth_s_eq),
+    finish [(gcf.continuants_aux_recurrence nth_s_eq ppconts_eq pconts_eq)] },
+  have : 0 ≤ ppconts.b, by
+  { have not_terminated_at_pred_n : ¬g.terminated_at (n - 1), by
+    { have : ¬g.terminated_at n, by { have : g.s.nth n ≠ none, by simp [nth_s_eq], exact this },
+      exact (mt (terminated_stable n.pred_le) this) },
+    have : (fib n : K) ≤ ppconts.b, by
+    { have : ¬g.terminated_at (n - 2), from
+        mt (terminated_stable (n - 1).pred_le) not_terminated_at_pred_n,
+      exact (of_fib_le_continuants_aux_b $ or.inr this) },
+    exact le_trans (by exact_mod_cast (fib n).zero_le) this },
+  solve_by_elim [le_add_of_nonneg_of_le, le_refl]
+end
+
+/-- Shows that `Bₙ₊₁ ≥ bₙ* Bₙ`, where `bₙ` is the `n`th partial denominator and `Bₙ₊₁` and `Bₙ` are
+the `n + 1`th and `n`th denominator of the continued fraction. -/
+theorem of_le_succ_nth_denom {b : K}
+  (nth_part_denom_eq : (gcf.of v).partial_denominators.nth n = some b) :
+  b * (gcf.of v).denominators n ≤ (gcf.of v).denominators (n + 1) :=
```

Comment:
I don't understand why this theorem is not trivial and why you need to work it through `of_le_succ_succ_nth_continuants_aux_b` whose proof involves Fibonacci numbers: you know that B_{n+1} = b_n * B_n + B_{n-1} and that B_{n-1} \geq 1 (except for the special case n = 0), so the inequality follows directly. Am I missing something? (or is it just the API unfolding that makes it harder than it looks?)

Reply:
We do not directly have `B_{n-1} >= 1` (or `>= 0`, which is sufficient here). One could prove this, of course, but it would be more or less the same proof as for `B_n >= fib (n + 1)`. So instead, I prove the stronger result `B_n >= fib (n + 1)` and then when I need `B_{n-1} >= 0` in your suggested approach above, I derive it using the facts `0 <= fib n` and `fib n <= B_{n - 1}`.

Reply:
ok, thanks for the explanation! It might be worth adding this as a comment somewhere.

Reply:
I added the comment together with the lemma that the denominators are nonnegative.